avinashi10/game-lobby-system
# Players Join & Leave Lobbies

## Description
This PR adds two player features. Players can join or leave lobbies and player lists will update for each lobby in real-time for all players. Also, when a player joins or leaves a lobby, all the other players in that lobby get notified directly in real time.


## Acceptance Criteria

- [x] Allow players to join or leave a lobby.
- [x] Notify all players in a lobby of players joining/leaving in real-time.




## What type of PR is this? (check all applicable)


|     | Type                       |
| --- | -------------------------- |
| ✓ | :sparkles: New feature     |
|     | 🎨 Style                   |
| ✓ | :hammer: Refactoring       |
|     | :bug: Bug fix              |
|   | :link: Update dependencies |



## Mobile & Desktop Screenshots/Recordings

<img width="1080" alt="Screenshot 2024-03-31 at 9 03 52 PM" src="https://github.com/avinashi10/game-lobby-system/assets/115492619/43b2be2c-943e-432a-88a2-889ba40dcf1b">


<img width="1310" alt="Screenshot 2024-03-31 at 9 04 50 PM" src="https://github.com/avinashi10/game-lobby-system/assets/115492619/c544290e-c530-419e-a74b-f5a446c47d9a">

## What I learned
I learned a lot more about setting up web sockets, specifically: `socket.to().emit` sends messages to all in a room except the sender, while `io.to().emit` includes the sender

## Testing steps

1.  Clone the repo locally and pull latest version.
2. Switch to the `feature/add-players` branch
3. To start the client and server, run `npm start` 
4. Open your browser to the port indicated when starting the client, open 3+ browser windows at this port to emulate several players
5. In each window create a distinct player, create several lobbies. Add and remove players from lobbies across all windows and ensure that player lists update accurately and that alerts appear only in relevant windows (for players in same lobby where action just happened).


## What gif best describes this PR?
  ![gif name](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExZjVpM3hwNGxud3N0OTUwMTY3N3Fib2F2ems2bmhvNGpzOGd2bDVwMCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/TdfyKrN7HGTIY/giphy.gif)
